### ISSUE TITLE
StatefulSet Pods do not have Cluster DNS

### ISSUE THREAD

**Issue by lizardruss** [kubedb/project#664]
Hello! Thanks for the great work on this project!

I have a feature request that would make connecting to a Redis Cluster easier, and would allow some of your [documentation](https://kubedb.com/docs/0.12.0/guides/redis/clustering/redis-cluster/) to use DNS names instead of IPs when connecting from within the same Kubernetes Cluster.

In order to connect to Redis running in Cluster mode, I should pass all the "Redis Nodes" into my client ([ioredis](https://github.com/luin/ioredis) in this case), and it's smart enough to connect to masters when writing, and can read from any of the nodes (master or slave).

In order to pass each "Redis Node" into my client, I either need to have an IP, or a DNS name for each node. With the current release of KubeDB, only the IP option is available because of the way the StatefulSet is created. [This issue](https://github.com/kubernetes/kubernetes/issues/74950#issuecomment-472285328) explains that each Pod of a StatefulSet only gets a DNS entry created if the `serviceName` of the StatefulSet matches a Service that targets the StatefulSet's pods.

KubeDB is already creating a Service for the StatefulSet of the Redis cluster, however its `serviceName` is set to `kubedb` instead of the name of its service. Changing [this line](https://github.com/kubedb/operator/blob/master/vendor/kubedb.dev/redis/pkg/controller/statefulset.go#L222) to:
```go
in.Spec.ServiceName = redis.ServiceName()
``` 
_should_ allow per-pod DNS names to be used.

With that change, each StatefulSet Pod would have a DNS record with the format `[pod-name].[service-name].[namespace].svc`. This would enable me to pass the DNS names to my client without needing to do the work of looking up each IP and reconfiguring my client if a pod crashes, or a slave becomes a master, etc.

I would need to know my Redis Cluster's topology to predict what the DNS names would be. For example, a cluster named `redis-cluster` with 3 masters and 1 replicas, running in namespace `foo`, I could assume the DNS names would be:
- `redis-cluster-shard0-0.redis-cluster.foo.svc`
- `redis-cluster-shard0-1.redis-cluster.foo.svc`
- `redis-cluster-shard1-0.redis-cluster.foo.svc`
- `redis-cluster-shard1-1.redis-cluster.foo.svc`
- `redis-cluster-shard2-0.redis-cluster.foo.svc`
- `redis-cluster-shard2-1.redis-cluster.foo.svc`

I've seen the suggestion to use `voyager` to connect, however running an ingress controller seems overkill when connecting from within the Kubernetes cluster and ClusterDNS could be used instead.

This change could also help any KubeDB integration that uses a StatefulSet and clients need connect to each node from within the Kubernetes Cluster. Perhaps #422 as well.

I'm happy to work on a PR for this if it all makes sense and seems like a good idea.

**Comment by shudipta** [comment]
Yeah, that was a shortcoming for Redis. If your client [ioredis] (https://github.com/luin/ioredis) can do things easier than builtin redis load-balancer, then you are welcome to contribute.

But one thing is wrong you suggested above. Changing [this line](https://github.com/kubedb/operator/blob/master/vendor/kubedb.dev/redis/pkg/controller/statefulset.go#L222) to

```go
in.Spec.ServiceName = redis.ServiceName()
```

is not necessary. Because here you have to use a headless service for each of the StatefulSet and we use the default headless (governing) service the operator creates named `kubedb`. Moreover, if you plan to use `redis.ServiceName()` which is database service and the `.spec.ClusterIP` is should be `None` in a headless service. That is why you can't do this. In that case, you can assume the DNS names are:

- `redis-cluster-shard0-0.kubedb.foo.svc`
- `redis-cluster-shard0-1.kubedb.foo.svc`
- `redis-cluster-shard1-0.kubedb.foo.svc`
- `redis-cluster-shard1-1.kubedb.foo.svc`
- `redis-cluster-shard2-0.kubedb.foo.svc`
- `redis-cluster-shard2-1.kubedb.foo.svc`

**Comment by shudipta** [comment]
Currently, we are focusing on something other pieces of stuff in KubeDB. So we need time to meet these shortcomings you wrote. We knew those but due to time and resource shortage, we were not able to settle these things. But we welcome the community to contribute.

Some points you are missing I suspect.

- The reason for `redis-cluster-shard0-0.kubedb.foo.svc` does not work is that the service `kubedb` does not have any selector in `.spec.selector`. Because this is a general headless service throughout the KubeDB project used for provisioning statefulsets of different kinds of databases. That's why it has no selector. But if needs, we create a new service for the purpose of headless service, like it is done in `kubedb/mysql`, `kubedb/mongodb`. If a headless service has no selector, one can not resolve the pods of statefulsets through it.

- So, the main point is the service you want to create must have selector(s) in your case.

- No need to check `redis.Mode == RedisModeCluster` condition. You can simply do it for both `Standalone` and `Cluster` mode.

- We did not create a new service for headless service. The reason was,
While working, we found that Redis can not resolve the DNS names of redis-servers. That is why we were not concerned about using a headless service having selectors. And for the same reason, we used IP instead of DNS names everywhere in the docs too.

I hope, this will find you well. So what do you think?

**Comment by lizardruss** [comment]
I looked over the `kubedb/mongo` services that are created. Perhaps you could help me understand why there is a [governing service](https://github.com/kubedb/mongodb/blob/master/pkg/controller/mongodb.go#L57) and a [database service](https://github.com/kubedb/mongodb/blob/master/pkg/controller/mongodb.go#L69). It looks like `kubedb/mongo` has replaced the `kubedb` service with a mongo specific `[mongo-db-name]-gvr` service.

I have been thinking that a small change to the 'database service' would be sufficient for my needs. If I were to change the creation of the stateful set to use the 'database service' name instead of the 'governing service' then perhaps we could remove the 'governing service'.

I don't want to do that if it will break something else, however. Is there a purpose for the `kubedb` service other than bootstrapping the stateful set?

**Comment by shudipta** [comment]
You are right.

- Where we need, we replace `kubedb` service with a new service `<database_crd_name>-gvr` within the same namespace as the CRD.

- In addition, changes to the database service named `<CRD_name>` would be sufficient for your needs. In that case, you do not need a governing service anymore. But KubeDB keeps both governing service and database service. Cause, the database statefulset need a headless service (here we use either `kubedb` service or the governing service for it whose `ClusterIP` is set to `None`). And the database service is used for backend database pods selected by its selector. Neither the `kubedb` service nor the governing service has a ClusterIP of its own.
But the database service has its own ClusterIP whose aim is to serve the requests sent by the clients to the backend servers. That is why we keep both of them.

- However, `kubedb` service has no other purpose other than bootstrapping the statefulsets. But when it needs, we create a new governing service named `<database_crd_name>-gvr` with `ClusterIP` `None`.

**Comment by pavel-khritonenko** [comment]
>     * We did not create a new service for headless service.

It might have the sense to fix the documentation:

https://kubedb.com/docs/v0.13.0-rc.0/guides/redis/clustering/redis-cluster/

> KubeDB operator will also create a governing service for StatefulSets with the name `<redis-name>-gvr`.

Because it doesn't create such service.

**Comment by vadasambar** [comment]
Hey @shudipta , I can work on this issue if you don't mind.  cc: @the-redback

**Comment by vadasambar** [comment]
Based on the discussion and my analysis of the code, redis controller is not creating a `<cr-name>-gvr` service right now like mongodb. StatefulSets need a headless service which resolves to the StatefulSet pods for pod DNS to work according to the official docs (check https://github.com/kubernetes/kubernetes/issues/74950#issuecomment-685872415).

**Comment by vadasambar** [comment]
Can someone please take a look at the above PR?

@shudipta , @the-redback , @tamalsaha

**Comment by vadasambar** [comment]
Above fix should work for both single node and multi-node redis cluster. I have tested it manually on my local kubernetes cluster.

**Comment by shudipta** [comment]
Sounds good 😊
